### PR TITLE
Optimize GatherElements and Scaler.

### PR DESCRIPTION
**Pull request by pranavsharma**
**Description**: Optimize GatherElements and Scaler. 
* Avoid unnecessary copy of the indices tensor in GatherElements. In this model, the indices tensor has 500k-1M elements.
* Parallelize Scaler.

Together these changes result in 37% reduction in runtime.

**Motivation and Context**
Changes done as part of investigating a ML.NET model.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc [hunk 1]
```diff
@@ -173,17 +159,24 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
...
 
       // process 1 chunk of 'inner dimension' length
-      for (int64_t i = 0; i < inner_dim_size; ++i) {
-        // optimizer will remove the redundant if/else block based on 'is_string' template parameter
-        if (is_string) {
-          output_data[++output_counter] = input_data[base_offset + (indices_data[++indices_counter] * input_shape_pitches[axis]) + i];
-        } else {
+      if (is_string) {
+        for (int64_t i = 0; i < inner_dim_size; ++i) {
+          // optimizer will remove the redundant if/else block based on 'is_string' template parameter
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/gather_elements.cc:164]: nit: should the comment be kept closer to the if...else (now that if and for are inter-changed). ? Same for the other 2 instances below...

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc [hunk 2]
```diff
@@ -154,15 +127,28 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
   const int64_t input_rank = static_cast<int64_t>(input_tensor->Shape().NumDimensions());
   const TensorPitches input_shape_pitches(*input_tensor);
 
-  const std::vector<int64_t>& indices_data = parse_and_validate_indices_tensor(indices_tensor, axis, input_tensor->Shape());
+  //const std::vector<int64_t>& indices_data = parse_and_validate_indices_tensor(indices_tensor, axis, input_tensor->Shape());
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/gather_elements.cc:130]: nit: can remove the commented line altogether now that we don't need the "corrected" indices in a container ?

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc [hunk 3]
```diff
@@ -193,13 +186,23 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
...
+      if (is_string) {
+        for (int64_t i = 0; i < inner_dim_size; ++i) {
+          // for innermost axis, input_shape_pitches[axis] = 1 (so no need to multiply)
+          // optimizer will remove the redundant if/else block based on 'is_string' template parameter
+          output_data[++output_counter] =
+              input_data[base_offset +
+                         GetNegativeIndexAdjustedValue<Tin>(indices_data, ++indices_counter, axis, input_shape)];
+        }
+      } else {
+        for (int64_t i = 0; i < inner_dim_size; ++i) {
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/gather_elements.cc:198]: For 500k - 1M indices, we can potentially get more gains by parallelizing this loop ?

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc [hunk 4]
```diff
@@ -86,42 +86,15 @@ static inline void increment_over_inner_dim(std::vector<int64_t>& current_dims,
...
-  if (indices_tensor->IsDataType<int32_t>()) {
-    const auto* data = indices_tensor->Data<int32_t>();
-    for (int64_t i = 0; i < num_elements; ++i)
-      indices_data.push_back(data[i]);
-  } else if (indices_tensor->IsDataType<int64_t>()) {
-    const auto* data = indices_tensor->Data<int64_t>();
-    for (int64_t i = 0; i < num_elements; ++i)
-      indices_data.push_back(data[i]);
+template <typename Tin>
+static int64_t GetNegativeIndexAdjustedValue(const Tin* indices_data, Tin index, int64_t axis, const TensorShape& input_shape) {
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/gather_elements.cc:90]: maybe we can provide the hint to the compiler that this method can be inlined ?

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc (hunk 3 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/tensor/gather_elements.cc:198]: Yeah, parallelizing this loop is pending. It'll take a little bit more work. I wanted to checkin first what I had working.

**Review by hariharans29 — APPROVED** [review]

**Pull request merged**